src/uct/ib/efa/ib_efa_md.c

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <uct/ib/efa/ib_efa.h>
+#include <ucs/type/status.h>
+#include <ucs/debug/log.h>
+
+
+static uct_ib_md_ops_t uct_ib_efa_md_ops;
+
+static ucs_status_t uct_ib_efa_md_open(struct ibv_device *ibv_device,
+                                       const uct_ib_md_config_t *md_config,
+                                       uct_ib_md_t **p_md)
+{
+    struct ibv_context *ctx;
+    uct_ib_efadv_md_t *md;
+    struct efadv_device_attr attr;
+    ucs_status_t status;
+    int ret;
+
+    if (md_config->devx == UCS_YES) {
+        ucs_error("DEVX requested but not supported");
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    ctx = ibv_open_device(ibv_device);
+    if (ctx == NULL) {
+        ucs_diag("ibv_open_device(%s) failed: %m",
+                 ibv_get_device_name(ibv_device));
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ret = efadv_query_device(ctx, &attr, sizeof(attr));
+    if (ret != 0) {
+        ucs_debug("efadv_query_device(%s) failed: %d",
+                  ibv_get_device_name(ibv_device), ret);
+        status = UCS_ERR_UNSUPPORTED;
+        goto err_free_context;
+    }
+
+    md = ucs_derived_of(uct_ib_md_alloc(sizeof(*md), "ib_efadv_md", ctx),
+                        uct_ib_efadv_md_t);
+    if (md == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_free_context;
+    }
+
+    md->super.super.ops = &uct_ib_efa_md_ops.super;
+    md->super.name      = UCT_IB_MD_NAME(efa);
+    md->efadv.attr      = attr;
+
+    status = uct_ib_device_query(&md->super.dev, ibv_device);
+    if (status != UCS_OK) {
+        goto err_md_free;
+    }
+
+    uct_ib_device_configure(&md->super.dev);
+
+    md->super.dev.mr_access_flags = IBV_ACCESS_LOCAL_WRITE;
+    if (uct_ib_efadv_has_rdma_read(&md->efadv)) {
+        md->super.dev.mr_access_flags |= IBV_ACCESS_REMOTE_READ;
+    }
+
+    md->super.dev.max_sq_sge            = md->efadv.attr.max_sq_sge;
+    md->super.dev.max_inline_data       = md->efadv.attr.inline_buf_size;
+    md->super.dev.ordered_send_comp     = 0;
+    md->super.dev.req_notify_cq_support = 0;
```

Comment:
fixed first point, second point, max_sge: removed, max_inline_data: I probably did not get it.

req_notify_cq_support:
- `IBV_DEVICE_N_NOTIFY_CQ`: from `ibv_devinfo -v` it is not set in `device_cap_flags` neither on aws/rock, so I think we cannot use it to differentiate (might be not directly related).
- Re-reading rmda-core code, I see `ibv_create_cq(.. , channel, ..)` will indeed return `EOPNOTSUP` when there is no completion support, but:
  - solicited_only is not supported: `efa_arm_cq()` in rdma-core returns `EOPNOTSUP` and I see ` ibv_req_notify_cq(rdmap200s0:1/IB, 0, sol=1) failed: Operation not supported `
  - there seems to be an issue where we end-up in ENOMEM situation with channel_create/create_cq/destroy_channel, I created issue https://github.com/amzn/amzn-drivers/issues/306 as it seems to me it could be in driver.

Based on that, if we want to add req_notify_cq_support event though there is not solicited_only capability, I would suggest to add it in a future PR.

Reply:
1- I do not see easy solution. One way would be to create/instantiate separate ops with only iface_query function different, but both seem overkill:
  - `uct_iface_ops_t uct_ud_verbs_iface_tl_ops` -> `uct_iface_ops_t uct_ud_verbs_efa_iface_tl_ops`
  - or adding `uct_ud_iface_ops_t uct_ud_verbs_efa_iface_ops` maybe iface_query along with existing create_qp

2- Yes. Channel apparently might also not be supported, seems it is, on my node but worth noting that channel creation/deletion is causing issues at least with our gtest, created issue mentioned. If we go that route, I would suggest reworking event to operate without reading with solicited_only=1 in a separate PR. Also, not sure if that's what you meant but I tried allowing `IBV_SEND_SOLICITED` regardless at sending side, but the posting now fails with 22, see below.

```
[ RUN      ] ud_verbs/test_many2one_am.am_bcopy/0 <ud_verbs/rdmap79s0:1>
[pool0-1400:1609659:0:1609659]    ud_verbs.c:102  Assertion `ret == 0' failed: ibv_post_send() returned 22 (Operation not supported)

(gdb) p wr->send_flags & IBV_SEND_SOLICITED
$1 = 4
```

Reply:
let's discuss offline how we can avoid adding some of these fields